src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/Tensor.cs

```diff
@@ -0,0 +1,355 @@
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualBasic;
+using static System.Runtime.InteropServices.JavaScript.JSType;
+
+#pragma warning disable CS8601 // Possible null reference assignment.
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+#pragma warning disable 8500 // address / sizeof of managed types
+
+namespace System.Numerics.Tensors
+{
+    public sealed class Tensor<T>
+        : IDisposable,
+          ITensor<Tensor<T>, T>
+        where T : IEquatable<T>
+    {
+        /// <summary>A byref or a native ptr.</summary>
+        internal readonly T[] _values;
+        /// <summary>The number of elements this Tensor contains.</summary>
+        internal readonly nint _linearLength;
+        /// <summary>The lengths of each dimension.</summary>
+        internal readonly nint[] _lengths;
+        internal readonly nint[] _strides;
+        /// <summary>If the backing memory is permanently pinned (so not just using a fixed statement).</summary>
+        internal readonly bool _isPinned;
+
+        /// <summary>
+        /// Creates a new empty Tensor.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Tensor()
+        {
+            _linearLength = 0;
+            _values = [];
+            _lengths = [];
+            _strides = [];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Tensor(T[] values, nint[] lengths, bool isPinned)
+        {
+            _linearLength = SpanHelpers.CalculateTotalLength(ref lengths);
+
+            Debug.Assert(_linearLength >= 0);
+
+            _values = values;
+            _lengths = lengths;
+            _strides = SpanHelpers.CalculateStrides(Rank, _lengths);
+            _isPinned = isPinned;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Tensor(T[] values, nint[] lengths, nint[] strides, bool isPinned)
+        {
+            _linearLength = SpanHelpers.CalculateTotalLength(ref lengths);
+
+            Debug.Assert(_linearLength >= 0);
+
+            _values = values;
+            _lengths = lengths;
+            _strides = strides;
+            if (strides == Array.Empty<nint>())
+                _strides = SpanHelpers.CalculateStrides(Rank, lengths);
+            _isPinned = isPinned;
+
+        }
+
+        // IDisposable
+
+        public void Dispose()
+        {
+        }
+
+        private static class EmptyTensor
+        {
+#pragma warning disable CA1825 // this is the implementation of Tensor.Empty<T>()
+            internal static readonly Tensor<T> Value = new Tensor<T>();
+#pragma warning restore CA1825
+        }
+        // ITensor
+
+        public static Tensor<T> Empty => EmptyTensor.Value;
+
+        public bool IsEmpty { get { return _lengths.Length == 0; } }
```

Comment:
If a tensor contains all zero lengths, is that not empty?

Reply:
An all zero length tensor is empty. I don't think I follow though. The idea behind the `Empty` is to allow comparing with `Tensor<T>.Empty`.

Reply:
`Empty` is fine. I'm asking about `IsEmpty`. I'd expect for example a Tensor<T> created around `Array.Empty<T>()`, which would have a single length of length 0, to have IsEmpty return true, but the implementation above will return false.